driver/compile_common.ml

```diff
@@ -37,14 +37,32 @@ let with_info ~native ~tool_name ~dump_ext unit_info k =
     native;
   }
 
+module Parse_result = struct
+  type 'a t = { ast : 'a; info : info }
+
+  let update_unit_info ~info ast =
+    let source_file = !Location.input_name in
+    let new_target =
+      Unit_info.update_human_source_file_name info.target source_file
+    in
+    { ast; info = { info with target = new_target } }
+
+  let map_ast { ast; info } ~f = { ast = f ast info; info }
```

Comment:
An idea to consider when you fix the testsuite (but you can also totally ignore it): rather than the generic `map_ast`, it could make sense to define a `print_if` function to specialize the function to its current uses:
```
|> Parse_result.printf_if Clflags.dump_parsetree Printast.interface
|> Parse_result.printf_if Clflags.dump_source Pprintast.signature
```

Reply:
Oh yeah I like this idea, thanks. I've done so.